front/src/applications/operationalStudies/components/MacroEditor/OsrdNgeSync.tsx

```diff
@@ -0,0 +1,918 @@
+  HaltezeitIPV: { haltezeit: 0, no_halt: false },
+  HaltezeitA: { haltezeit: 0, no_halt: false },
+  HaltezeitB: { haltezeit: 0, no_halt: false },
+  HaltezeitC: { haltezeit: 0, no_halt: false },
+  HaltezeitD: { haltezeit: 0, no_halt: false },
+  HaltezeitUncategorized: { haltezeit: 0, no_halt: false },
+};
+
+const DEFAULT_LABEL_GROUP: LabelGroupDto = {
+  id: 1,
+  name: 'Default',
+  labelRef: 'Trainrun',
+};
+
+const DEFAULT_TRAINRUN_CATEGORY: TrainrunCategory = {
+  id: 1, // In NGE, Trainrun.DEFAULT_TRAINRUN_CATEGORY
+  order: 0,
+  name: 'Default',
+  shortName: '', // TODO: find a better way to hide this in the graph
+  fachCategory: 'HaltezeitUncategorized',
+  colorRef: 'EC',
+  minimalTurnaroundTime: 0,
+  nodeHeadwayStop: 0,
+  nodeHeadwayNonStop: 0,
+  sectionHeadway: 0,
+};
+
+const DEFAULT_TRAINRUN_TIME_CATEGORY: TrainrunTimeCategory = {
+  id: 0, // In NGE, Trainrun.DEFAULT_TRAINRUN_TIME_CATEGORY
+  order: 0,
+  name: 'Default',
+  shortName: '7/24',
+  dayTimeInterval: [],
+  weekday: [1, 2, 3, 4, 5, 6, 7],
+  linePatternRef: '7/24',
+};
+
+const DEFAULT_DTO: NetzgrafikDto = {
+  resources: [],
+  nodes: [],
+  trainruns: [],
+  trainrunSections: [],
+  metadata: {
+    netzgrafikColors: [],
+    trainrunCategories: [DEFAULT_TRAINRUN_CATEGORY],
+    trainrunFrequencies: [...DEFAULT_TRAINRUN_FREQUENCIES],
+    trainrunTimeCategories: [DEFAULT_TRAINRUN_TIME_CATEGORY],
+  },
+  freeFloatingTexts: [],
+  labels: [],
+  labelGroups: [],
+  filterData: {
+    filterSettings: [],
+  },
+};
+
+const DEFAULT_TIME_LOCK: TimeLockDto = {
+  time: null,
+  consecutiveTime: null,
+  lock: false,
+  warning: null,
+  timeFormatter: null,
+};
+
+type NodeIndex = { node: MacroNodeResponse; saved: boolean };
+export default class OsrdNgeSync {
+  /**
+   * Storing nodes by path item key
+   * It's the main storage for node.
+   * The saved attribut is to know if the data comes from the API
+   * If the value is a string, it's a key redirection
+   */
+  private nodesByPathKey: Record<string, NodeIndex | string>;
+
+  /**
+   * We keep a dictionnary of id/key to be able to find a node by its id
+   */
+  private nodesIdToKey: Record<number, string>;
+
+  /**
+   * Storing labels
+   */
+  private labels: Set<string>;
+
+  /**
+   * NGE resource
+   */
+  private ngeResource = { id: 1, capacity: 0 };
+
+  /**
+   * Default constructor
+   */
+  constructor(
+    private dispatch: AppDispatch,
+    private readonly scenario: ScenarioResponse,
+    private trainSchedules: TrainScheduleResult[]
+  ) {
+    // Empty
+    this.labels = new Set<string>([]);
+    this.nodesIdToKey = {};
+    this.nodesByPathKey = {};
+    this.ngeResource = { id: 1, capacity: trainSchedules.length };
+  }
+
+  /**
+   * Load & index the data of the train schedule for the given scenario
+   */
+  async loadAndIndex(): Promise<void> {
+    // Load path items
+    this.trainSchedules
+      .map((schedule) => schedule.path)
+      .flat()
```

Comment:
Why not directly a flatMap ?
```suggestion
      .flatMap((schedule) => schedule.path)
```